nupic/research/frameworks/vernon/README.md

```diff
@@ -71,16 +74,28 @@ class UpdateBoostStrength:
         return eo
     pass
 ```
-one can see that **update_boost_strength** will be applied in the parent class's pre_epoch period. 
+one can see that **update_boost_strength** will be applied in the parent class's pre_epoch period.
+
+Mixins have full access to the experiment class, and they can do whatever they want. Often, they will extend methods, first calling `super().method` then running additional code. Other times, they will replace methods (never calling `super().method`). Mixins can depend on the internals of specific experiments and even the internals of other mixins. Because this is research code, we don't place any firm restrictions on how mixins should work. To make the dependencies explicit and to promote reusability, as mixins become mature they should use the following practices:
 
-**get_execution_order** tracks the use of mixins at various stages of the experiment, and the execution order list is displayed when an experiment is run. 
+- Interact with the `self` via explicit **interfaces**. Rather than having mixins interact directly with specific experiments, the experiment should implement an [interface](./interfaces/), and the mixin should interact with that interface. This enables using the mixin with any experiment exposing this interface. It also prevents errors; if any mixins have an unresolved interface dependency, the experiment class will have unimplemented abstract methods and will fail to instantiate. Finally, this enables static analysis and IDE support.
+- Implement the `get_execution_order` classmethod. This documents the mixin's modifications to various stages of the experiment. The execution order list is displayed when an experiment is run.
 
 Vernon provides a variety of mixins that enable the specification of more advanced experiments, including:
 - custom loss functions (composite_loss.py)
 - knowledge distillation to have the network learn from a teacher model (knowledge_distillation.py)
 - efficient tuning of the learning rate (lr_range_test.py)
 - MaxUp data augmentation to improve model generalization (maxup.py)
 
+Interface edge cases
```

Comment:
Maybe we can add links that give an overview of interfaces and mixins. These are new concepts for me and I imagine they would be new for others even if they're familiar with OOP.

Here are some links we could consider adding:
mixins - https://realpython.com/inheritance-composition-python/#mixing-features-with-mixin-classes
interfaces - https://realpython.com/python-interface/#python-interface-overview